main/resource-access.bs

```diff
@@ -50,6 +50,51 @@ A Solid server MUST reject `PUT`, `POST` and `PATCH` requests without the
 `Content-Type` header with a status code of `400`.
 [[Source](https://github.com/solid/specification/issues/70#issuecomment-535499628)]
 
+#### Deleting resources #### {#delete}
+
+Servers MUST support the HTTP `DELETE` method [[!RFC7231]].
+[[Source](https://github.com/solid/specification/issues/39#issuecomment-538017667)]
+
+When using Web Access Control ([[#wac]]):
+
+Servers MUST NOT allow delete requests to a storage's root container and associated ACL resource and exclude the `DELETE` method in the HTTP response header `Allow` in response to safe method requests [[!RFC7231]].
+[[Source](https://github.com/solid/specification/issues/37#issuecomment-627281466)]
+
+To delete a resource, an `acl:agent` MUST have `acl:Write` privilege per the ACL inheritance algorithm on the resource and the containing container.
+[[Source](https://github.com/solid/solid-spec/issues/195#issuecomment-559799154)]
+
+When a contained resource is deleted, the server MUST also remove the corresponding containment triple, which has the effect of removing the deleted resource from the containing container.
+[[Source](https://www.w3.org/TR/ldp#ldpc-del-contremovesconttriple)]
+
+When a contained resource is deleted, the server MUST also delete the associated resources (see the [[#rm]] section).
+
+To delete an ACL resource (see [[#ar-wac]]), an `acl:agent` MUST have `acl:Control` privileges per the ACL inheritance algorithm on the resource directly associated with it.
+[[Source](https://github.com/solid/specification/issues/145)]
+[[Source](https://github.com/solid/specification/issues/41)]
+
+A container can only be deleted if it contains no resources ([[#resource-containment]]). When a `DELETE` method request is made to a container, the server MUST delete the contained resources except containers. To recursively delete a container, server MUST accept client's `DELETE` request including the HTTP `Prefer` header with `return="representation"; include="http://www.w3.org/ns/ldp#PreferContainment"` (see [[!LDP]]'s Preferences on the Prefer Request Header). Server MUST respond with the `409` status code and response body including containment triples about unaffected resources of the delete request for agents with `acl:Read` privilege per the ACL inheritance algorithm.
```

Comment:
This is confusing. First, this sentence states that a container _can only be deleted if it contains no resources_. Then it describes recursive delete, which contradicts the first statement.

As a general comment, recursive delete becomes _very_ complicated in the case of complex ACLs, atomicity guarantees and the prospect of partial failure. If you plan to support recursive delete in the specification, I would suggest that there be a mechanism for reporting partial failures to clients (or, as would be my preference, don't support recursive delete at all)

Reply:
Also, the `Prefer: return=representation` part is strange. That hint (`return=representation`) is typically used to control the _response_ of the HTTP operation, not the internal server behavior.

Reply:
Well, I assume the first sentence could still hold if you implement recursive delete via a depth-first search - i.e. the server starts deleting the leaf-nodes, and then only deletes child containers once they are empty, and keep popping back up the stack.
Partial failures do still make be nervous though - i.e. having some mechanism to report them back to the user, and making it very clear to client-app developers to **_not_** assume a recursive delete **_will just happen_** on the server (i.e. where they update their user interface to remove an entire folder structure). Instead they need to be aware that the delete operation might partially fail (and if it does fail for one resource, does it stop the recursive delete operation at that point, or does it continue deleting remaining leaf-nodes and remaining empty containers?). And so those developers need to know that they need to wait for a server response (or an async notification) telling them what actually happened on the server before updating their UI.

Reply:
I see what you mean with the confusion. It was intended to cover the DELETE request to a container. Perhaps it is redundant in that the third sentence mentions recursion - with the assumption that includes the effective request URI (ie. the container itself), in addition to the contained resources. If we omit the first sentence, it is clear?

Reply:
True re: `Prefer: return=representation` and typical use so far. I figured that Prefer header in context of DELETE would be sufficient. It still has the client preferring to see the resulting representation of the container. This is where the 409 also comes into play ie. if there is a partial fail, client can know (provided that they have Read).

There wasn't a particular way where a client can request its preference to delete a container recursively - at least in the issues that I've looked into. Happy to look at alternative approaches.

Reply:
The rough consensus was that the base DELETE would only remove the non-container resources, and to revisit recursive deletion. Prefer may be one way to do this.

Does Trellis support recursive delete? How would Trellis respond to `DELETE /foo/` , where `/foo/` contains `/foo/bar/` and containing `/foo/bar/baz`?

Reply:
No, Trellis does not support recursive delete. Let me emphasize that recursive delete is _hugely_ problematic.

Imagine the case of a container with 100,000 child resources. What happens if the HTTP connection times out before the delete operation finishes? Or if a user clicks a "stop" button in the browser?

Right now, Trellis would delete the container and leave an orphan nodes at /foo/bar/. I have some code, however, that would case the request to fail with a 409 error, preventing clients from deleting non-empty containers.

Reply:
I have considered the `Expect` header but decided against it. Would reusing LDP's Prefer pattern be completely wrong in context of DELETE?

Reply:
I would expect servers would have a approximation or limit the time in which they have to generate a response. We don't currently have a way (not to forget that there was no corresponding use case) for a client to send an additional request to stop the ongoing delete.

We can't have orphaned nodes in Solid ( https://github.com/solid/specification/blob/88baa6eb7bbd937f524150109bb8925ab39f619a/main/resource-access.bs#L103-L108 ).

Reply:
@pmcb55 Indeed eg `DELETE /foo/` would remove `/foo/` if it has no contained resources. If `/foo/` contains resources, `DELETE /foo/` would remove them instead. I don't particularly like how the operation is different depending on what the target resource contains, but I do see its appeal. In the PR's wording, as mentioned earlier, the first sentence probably goes better with the third sentence; recursive delete also removing the container itself (effective request URI).

Reply:
While server accepts the request for delete recursively, it could still respond with a 409 or another 4xx. Alternatively, instead of MUST, we use MAY (irrespective of the mechanism used to indicate client's wish for recursive delete).

Reply:
> DELETE /foo/ would remove /foo/ if it has no contained resources. If /foo/ contains resources, DELETE /foo/ would remove them instead.

FWIW, I would be strongly 👎  on this.

Imagine the case where you have two clients writing to `/foo/`. The first checks `/foo/` and finds that it is empty, it then issues `DELETE /foo/` expecting that it has deleted `/foo/`, but in fact it has deleted `/foo/bar` and `/foo/baz` (which were created by the second client the intervening time).

That scenario would become extremely difficult for clients to reason about.

Reply:
> Would reusing LDP's Prefer pattern be completely wrong in context of DELETE?

The use of `return=representation` is incorrect. The use of Prefer is not incorrect.

Reply:
Nods.. 

Proposal: postpone if and how recursive deletion can work to another PR. For now, the base requirement needs to be *along these lines*:

When a `DELETE` method request is made to a container, the server MUST delete the container if it contains no resources. If the container contains resources, the server MUST respond with the `409` status code and response body describing the error.

When we need to, we should add recursive delete in a way without interfering with those requirements.

Reply:
Happy to see a version with no MAY and no "prefer".  This spec has to be tight and unambiguous.  Any time the word MAY appeared is a red flag!  And using anything like "prefer" to actually have hard must semantics sounded a bad idea.

Reply:
Updated in https://github.com/solid/specification/pull/187/commits/c11eab495839bdaab5f51e06bc153c1260d51611